lib/intercept.js

```diff
@@ -235,6 +235,27 @@ function ErroringClientRequest(error) {
 
 inherits(ErroringClientRequest, http.ClientRequest)
 
+function urlToOptions(input) {
+  const options = {
+    protocol: input.protocol,
+    hostname: input.hostname.startsWith('[')
+      ? input.hostname.slice(1, -1)
+      : input.hostname,
+    hash: input.hash,
+    search: input.search,
+    pathname: input.pathname,
+    path: `${input.pathname}${input.search}`,
+    href: input.href,
+  }
+  if (input.port !== '') {
+    options.port = Number(input.port)
+  }
+  if (input.username || input.password) {
+    options.auth = `${input.username}:${input.password}`
+  }
+  return options
```

Comment:
It would be good to include tests to cover the rest of this function. It might be tempting to write these as unit tests, though we prefer to write functional tests via the public API because it helps ensure that all the code in the library is demanded by a real use case.

Reply:
Added some more tests, specifically, tests that use http.request and http.get, which is what others will actually use.